contracts/IdleTokenFungible.sol

```diff
@@ -0,0 +1,759 @@
+    returns (uint256 mintedTokens) {
+    _updateFeeInfo();
+
+    _minterBlock = keccak256(abi.encodePacked(tx.origin, block.number));
+    // Get current IdleToken price
+    uint256 idlePrice = _tokenPrice();
+    // transfer tokens to this contract
+    IERC20Detailed(token).safeTransferFrom(msg.sender, address(this), _amount);
+
+    mintedTokens = _amount * ONE_18 / idlePrice;
+    _mint(msg.sender, mintedTokens);
+
+    lastNAV += _amount;
+
+    if (_referral != address(0)) {
+      emit Referral(_amount, _referral);
+    }
+  }
+
+  /**
+   * Here we calc the pool share one can withdraw given the amount of IdleToken they want to burn
+   *
+   * @param _amount : amount of IdleTokens to be burned
+   * @return redeemedTokens : amount of underlying tokens redeemed
+   */
+  function redeemIdleToken(uint256 _amount)
+    external
+    returns (uint256) {
+      return _redeemIdleToken(_amount);
+  }
+
+  /**
+   * Here we calc the pool share one can withdraw given the amount of IdleToken they want to burn
+   *
+   * @param _amount : amount of IdleTokens to be burned
+   * @return redeemedTokens : amount of underlying tokens redeemed
+   */
+  function _redeemIdleToken(uint256 _amount)
+    internal nonReentrant
+    returns (uint256 redeemedTokens) {
+      _checkMintRedeemSameTx();
+      _updateFeeInfo();
+      if (_amount != 0) {
+        uint256 price = _tokenPrice();
+        uint256 valueToRedeem = _amount * price / ONE_18;
+        uint256 balanceUnderlying = _contractBalanceOf(token);
+
+        if (valueToRedeem > balanceUnderlying) {
+          redeemedTokens = _redeemHelper(_amount, balanceUnderlying);
+        } else {
+          redeemedTokens = valueToRedeem;
+        }
+        // update lastNAV
+        lastNAV -= redeemedTokens;
+        // burn idleTokens
+        _burn(msg.sender, _amount);
+        // send underlying minus fee to msg.sender
+        _transferTokens(token, msg.sender, redeemedTokens);
+      }
+  }
+
+  function _redeemHelper(uint256 _amount, uint256 _balanceUnderlying) private returns (uint256 redeemedTokens) {
+    address currToken;
+    uint256 idleSupply = totalSupply();
+    address[] memory _allAvailableTokens = allAvailableTokens;
+
+    for (uint256 i = 0; i < _allAvailableTokens.length; i++) {
+      currToken = _allAvailableTokens[i];
+      redeemedTokens += _redeemProtocolTokens(
+        currToken,
+        // _amount * protocolPoolBalance / idleSupply
+        _amount * _contractBalanceOf(currToken) / idleSupply // amount to redeem
+      );
+    }
+    // and get a portion of the eventual unlent balance
+    redeemedTokens += _amount * _balanceUnderlying / idleSupply;
+  }
+
+  /**
+   * Dynamic allocate all the pool across different lending protocols if needed,
+   * rebalance without params
+   *
+   * NOTE: this method can be paused
+   *
+   * @return : whether has rebalanced or not
+   */
+  function rebalance() external returns (bool) {
+    return _rebalance();
+  }
```

Comment:
just to confirm, is it ok that anyone can rebalance? i saw best-yield code.

Reply:
Yes, there are no incentives to call it atm, but is open because the method simply enforces the last allocations set. Allocations can be set only by rebalancer role